### PR TITLE
Avoid pointers in SpanConfig, TracerConfig, EventConfig

### LINKED ISSUE #2065: Avoid unnecessary heap allocation in NewSpanStartConfig
### Problem Statement

Right now calling `NewSpanStartConfig` will heap allocate `SpanConfig` every time is called.

### Proposed Solution

Multiple solutions can be implemented:
* Expose `Apply` in the Options interfaces (add another internal func to avoid extending it), so users will execute what `NewSpanStartConfig` does but with a `var sc SpanConfig` and pass the address to `sc`.
* Replace `NewSpanStartConfig` with a `ResolveSpanConfig` (or other suitable name) that accepts a pointer to SpanConfig so internally SpanConfig can be stack allocated and a pointer to sc can be passed to the resolve func.

Same applies to all configs.

### PR DESCRIPTION
**Pull request by jmacd**
Fixes #2065.

This does not appear to change any of the benchmarks in `sdk/trace`, however I would recommend this PR anyway because it feels more idiomatic to return structs, not pointers, for static configuration.

Before (at HEAD):
```
BenchmarkStartEndSpan/AlwaysSample-12    	 1348563	       876.4 ns/op	     848 B/op	       9 allocs/op
BenchmarkStartEndSpan/NeverSample-12     	 1633045	       734.8 ns/op	     752 B/op	       8 allocs/op
BenchmarkSpanWithAttributes_4/AlwaysSample-12         	  827178	      1465 ns/op	    1584 B/op	      17 allocs/op
BenchmarkSpanWithAttributes_4/NeverSample-12          	 1403071	       858.6 ns/op	     944 B/op	       9 allocs/op
BenchmarkSpanWithAttributes_8/AlwaysSample-12         	  599956	      1973 ns/op	    2112 B/op	      23 allocs/op
BenchmarkSpanWithAttributes_8/NeverSample-12          	 1271044	       938.9 ns/op	    1136 B/op	       9 allocs/op
BenchmarkSpanWithAttributes_all/AlwaysSample-12       	  643704	      1839 ns/op	    1936 B/op	      21 allocs/op
BenchmarkSpanWithAttributes_all/NeverSample-12        	 1319778	       907.7 ns/op	    1072 B/op	       9 allocs/op
BenchmarkSpanWithAttributes_all_2x/AlwaysSample-12    	  377712	      3129 ns/op	    3236 B/op	      32 allocs/op
BenchmarkSpanWithAttributes_all_2x/NeverSample-12     	 1000000	      1069 ns/op	    1392 B/op	       9 allocs/op
```

AFTER (this PR):
```
BenchmarkAttributesMapToKeyValue-12      	  452428	      2629 ns/op
BenchmarkStartEndSpan/AlwaysSample-12    	 1486172	       806.1 ns/op	     656 B/op	       7 allocs/op
BenchmarkStartEndSpan/NeverSample-12     	 1716672	       695.6 ns/op	     656 B/op	       7 allocs/op
BenchmarkSpanWithAttributes_4/AlwaysSample-12         	  857193	      1383 ns/op	    1392 B/op	      15 allocs/op
BenchmarkSpanWithAttributes_4/NeverSample-12          	 1473076	       812.7 ns/op	     848 B/op	       8 allocs/op
BenchmarkSpanWithAttributes_8/AlwaysSample-12         	  634714	      1882 ns/op	    1920 B/op	      21 allocs/op
BenchmarkSpanWithAttributes_8/NeverSample-12          	 1330555	       900.2 ns/op	    1040 B/op	       8 allocs/op
BenchmarkSpanWithAttributes_all/AlwaysSample-12       	  697209	      1723 ns/op	    1744 B/op	      19 allocs/op
BenchmarkSpanWithAttributes_all/NeverSample-12        	 1379874	       886.3 ns/op	     976 B/op	       8 allocs/op
BenchmarkSpanWithAttributes_all_2x/AlwaysSample-12    	  404030	      3066 ns/op	    3044 B/op	      30 allocs/op
BenchmarkSpanWithAttributes_all_2x/NeverSample-12     	 1000000	      1001 ns/op	    1296 B/op	       8 allocs/op
BenchmarkTraceID_DotString-12                         	17515768	        68.86 ns/op
BenchmarkSpanID_DotString-12                          	22704256	        53.28 ns/op
```

### CODE REVIEW

**Comment by MrAlias** [comment]
It would be helpful if we can show this change will reduce the number of allocation in a benchmark. The existing benchmarks show no change, but I suspect this is because they are not testing what is being changed.

@jmacd is this something you can do or would you like some help with this?

**Comment by jmacd** [comment]
I'd like to help. To take this further, I ran 

```
go test -bench=BenchmarkStartEndSpan/AlwaysSample -test.benchtime=10s -test.memprofile=mem.out
```

and looked at the output with

```
go tool pprof -alloc_objects -lines ./mem.out
```

The result after my previous PR shows 9 allocations, paraphrased as:

1. the new Context.Value
2. the start `SpanConfig` escapes
3. the end `SpanConfig` escapes
4. the new `*span` 
5. crap load of attribute map stuff (5 allocs)

I was able to modify this PR to produce the benchmark improvement. I'd like help completing this PR from someone in the group, if possible. The PR is mostly complete, but the same fix should be done in a bunch of places for consistency. `TracerConfig` and `EventConfig`.

The fix is a little ugly, too. (I've done this before, 🤷.)

The option functions can't take pointers, so you end up with Options like:

```
func (o timestampOption) applySpan(c SpanConfig) SpanConfig {
	c.timestamp = time.Time(o)
	return c
}
```

**Comment by jmacd** [comment]
Also, if we're looking for opportunities to improve allocations, the crap load of attribute map allocs are worth looking at.

image

**Comment by jmacd** [comment]
You can see the 7 resulting allocs in the picture above. (5 of them are the attribute map/list!)

**Comment by jmacd** [comment]
I updated the benchmark results, now showing improvement. 

```
BenchmarkAttributesMapToKeyValue-12      	  452428	      2629 ns/op
BenchmarkStartEndSpan/AlwaysSample-12    	 1486172	       806.1 ns/op	     656 B/op	       7 allocs/op
BenchmarkStartEndSpan/NeverSample-12     	 1716672	       695.6 ns/op	     656 B/op	       7 allocs/op
BenchmarkSpanWithAttributes_4/AlwaysSample-12         	  857193	      1383 ns/op	    1392 B/op	      15 allocs/op
BenchmarkSpanWithAttributes_4/NeverSample-12          	 1473076	       812.7 ns/op	     848 B/op	       8 allocs/op
BenchmarkSpanWithAttributes_8/AlwaysSample-12         	  634714	      1882 ns/op	    1920 B/op	      21 allocs/op
BenchmarkSpanWithAttributes_8/NeverSample-12          	 1330555	       900.2 ns/op	    1040 B/op	       8 allocs/op
BenchmarkSpanWithAttributes_all/AlwaysSample-12       	  697209	      1723 ns/op	    1744 B/op	      19 allocs/op
BenchmarkSpanWithAttributes_all/NeverSample-12        	 1379874	       886.3 ns/op	     976 B/op	       8 allocs/op
BenchmarkSpanWithAttributes_all_2x/AlwaysSample-12    	  404030	      3066 ns/op	    3044 B/op	      30 allocs/op
BenchmarkSpanWithAttributes_all_2x/NeverSample-12     	 1000000	      1001 ns/op	    1296 B/op	       8 allocs/op
BenchmarkTraceID_DotString-12                         	17515768	        68.86 ns/op
BenchmarkSpanID_DotString-12                          	22704256	        53.28 ns/op
```

File: trace/config.go [hunk 1]
```diff
@@ -28,20 +28,20 @@ type TracerConfig struct {
...
 }
 
 // NewTracerConfig applies all the options to a returned TracerConfig.
-func NewTracerConfig(options ...TracerOption) *TracerConfig {
-	config := new(TracerConfig)
+func NewTracerConfig(options ...TracerOption) TracerConfig {
+	var config TracerConfig
 	for _, option := range options {
-		option.apply(config)
+		option.apply(&config)
```
Comment (MrAlias) [inline trace/config.go:44]: Should this be updated to the alternate function signature, it takes and returns a struct instead of a pointer?

**Review by MrAlias — COMMENTED** [review]
If we are planning to change the `apply` method we need to update the contributing docs and like you pointed out make sure we are consistent across the board in how we apply things.

**Comment by Aneurysm9** [comment]
> The option functions can't take pointers, so you end up with Options like:
> 
> ```
> func (o timestampOption) applySpan(c SpanConfig) SpanConfig {
> 	c.timestamp = time.Time(o)
> 	return c
> }
> ```

I'm a bit confused by this.  How are the option functions not able to accept pointers and how does taking the config struct by value and returning a modified copy result in fewer allocations?

**Comment by MrAlias** [comment]
Superseded with https://github.com/open-telemetry/opentelemetry-go/pull/2212

The overall memory allocation issue will be addressed in a follow on to #2212 and will likely be based on the work found in this PR that will address #2065.

**Pull request closed without merging**